executor/src/witgen/range_constraints.rs

```diff
@@ -194,50 +194,69 @@ fn range_multiple<T: FieldElement>(min: T, max: T, factor: T) -> (T, T) {
     // We shift both intervals until they are both non-wrapping intervals.
     // If we do not succeed after shifting both of them by the smallest amount,
     // it means that the intersection cannot be expressed as a single interval.
     // In that case we just choose the smaller of the two inputs.
-    let shift: T = if a.0 <= a.1 {
-        // a is regular
-        if b.0 <= b.1 {
-            0.into()
-        } else if a.1 <= b.1 || b.1 <= a.0 {
-            b.1
-        } else if a.1 < b.0 {
-            a.1
-        } else {
-            // intersection consists of two intervals.
-            if range_width(a.0, a.1) >= range_width(b.0, b.1) {
-                return Some(a);
-            } else {
-                return Some(b);
-            }
-        }
-    } else if b.0 <= b.1 {
-        // a is wrapping b is regular
-        return interval_intersection(b, a);
+    // let shift: T = if a.0 <= a.1 {
+    //     // a is regular
+    //     if b.0 <= b.1 {
+    //         0.into()
+    //     } else if a.1 <= b.1 || b.1 <= a.0 {
+    //         b.1
+    //     } else if a.1 < b.0 {
+    //         a.1
+    //     } else {
+    //         // intersection consists of two intervals.
+    //         if range_width(a.0, a.1) >= range_width(b.0, b.1) {
+    //             return Some(a);
+    //         } else {
+    //             return Some(b);
+    //         }
+    //     }
+    // } else if b.0 <= b.1 {
+    //     // a is wrapping b is regular
+    //     return interval_intersection(b, a);
+    // } else {
+    //     // both a and b are wrapping
+    //     if cmp::max(a.1, b.1) < cmp::min(a.0, b.0) {
+    //         cmp::max(a.1, b.1)
+    //     } else {
+    //         // intersection consists of two intervals
+    //         if range_width(a.0, a.1) >= range_width(b.0, b.1) {
+    //             return Some(a);
+    //         } else {
+    //             return Some(b);
+    //         }
+    //     }
+    // };
+    if let Some((shift, (a_shifted, b_shifted))) =
+        [0.into(), a.0, a.1 + 1.into(), b.0, b.1 + 1.into()]
+            .into_iter()
+            .find_map(|shift| {
+                let a_shifted = shifted_interval(a, -shift);
+                let b_shifted = shifted_interval(b, -shift);
+                (a_shifted.0 <= a_shifted.1 && b_shifted.0 <= b_shifted.1)
+                    .then_some((shift, (a_shifted, b_shifted)))
+            })
+    {
+        let intersection = (
+            cmp::max(a_shifted.0, b_shifted.0),
+            cmp::min(a_shifted.1, b_shifted.1),
+        );
+        // If min is larger than max, the intersection is empty.
+        (intersection.0 <= intersection.1).then_some(shifted_interval(intersection, shift))
     } else {
-        // both a and b are wrapping
-        if cmp::max(a.1, b.1) < cmp::min(a.0, b.0) {
-            cmp::max(a.1, b.1)
+        // The intersection consists of two intervals
+        // TODO is this really true?
+        // Is the code at least good?
```

Comment:
Yes, I think this is true. The only way for none of the tested shifts to place 0 on a non-union interval is if the union is the whole field. In other words, if there is a non-union interval, it must start and end at two of the extremities of  a and b.